contracts/galactic-express/src/lib.rs

```diff
@@ -122,322 +131,292 @@ impl Contract {
+            return Err(Error::NotEnoughParticipants);
+        }
 
-        for (recipient, _) in scores.into_iter().rev() {
-            let amount = self.reward - deductible;
+        self.register(participant, false)?;
 
-            if FTokenEvent::Ok
-                != msg::send_for_reply_as(
-                    self.sft,
-                    FTokenAction::Message {
-                        transaction_id: stepper.step()?,
-                        payload: LogicAction::Mint { recipient, amount },
-                    },
-                    0,
-                    0,
-                )?
-                .await?
+        let mut random = Random::new()?;
+        let mut turns = vec![];
+
+        for turn_index in 0..TOTAL_TURNS {
+            let mut turn: HashMap<ActorId, Turn> = HashMap::new();
+
+            let first_turn: HashMap<ActorId, Turn> = self
+                .participants
+                .iter()
+                .map(|(actor, participant)| {
+                    (
+                        *actor,
+                        Turn::Alive {
+                            fuel_left: participant.fuel_amount,
+                            payload_amount: participant.payload_amount,
+                        },
+                    )
+                })
+                .collect();
+
+            let participants_data = if turn_index == 0 {
+                &first_turn
+            } else {
+                &turns[turn_index - 1]
+            };
+
+            for (actor, turn_entry) in participants_data
+                .into_iter()
+                .map(|(actor, turn_entry)| (*actor, *turn_entry))
             {
-                return Err(Error::TransferFailed);
-            }
-
-            if matches!(rankings.last(), Some((_, reward)) if *reward != amount) {
-                deductible += self.reward / 10 * 6 / PARTICIPANTS as u128;
+                let (fuel_left, payload_amount) = match turn_entry {
+                    Turn::Alive {
+                        fuel_left,
+                        payload_amount,
+                    } => (fuel_left, payload_amount),
+                    destroyed => {
+                        turn.insert(actor, destroyed);
+
+                        continue;
+                    }
+                };
+
+                turn.insert(
+                    actor,
+                    match self.turn(turn_index, &mut random, payload_amount, fuel_left) {
+                        Ok(new_fuel_left) => Turn::Alive {
+                            fuel_left: new_fuel_left,
+                            payload_amount,
+                        },
+                        Err(reason) => Turn::Destroyed(reason),
+                    },
+                );
             }
 
-            rankings.push((recipient, amount));
+            turns.push(turn);
         }
 
-        let mut io_turns: Vec<Vec<(ActorId, TurnOutcome)>> = vec![];
+        self.is_session_ended = true;
+        self.turns = turns.clone();
```

Comment:
unfortunately it won't be possible, because then we use the turn variable again to return `-> Result<Event, Error>`
```
Ok(Event::GameFinished(
    turns
        .into_iter()
        .map(|turn| turn.into_iter().collect())
        .collect(),
))
```